improved CDF based quantile estimator for q-digest

DanielTing opened the pull request:
This greatly improves the accuracy of the quantile estimator. The deterministic error guarantee is also preserved. The improvement works by using a good estimate of the cumulative distribution function (CDF) using linear interpolation. Existing implementations use the most extreme CDF that is consistent with the digest.

Commit:
improved CDF based quantile estimator for q-digest

electrum committed:
Use default thread count for TestingHttpServerModule

electrum committed:
Fix CHANGES

electrum committed:
Support prefixes in AbstractConfigurationAwareModule

electrum committed:
Fix behavior of BoundedThreadPool

Previously, it would behave as a fixed thread pool by
always creating up to the maximum number of threads.

electrum committed:
Use idle thread count for BoundedThreadPool

electrum committed:
Require at least one core thread for BoundedThreadPool

electrum committed:
Make TestBoundedThreadPool run single threaded

erichwang committed:
Strictly enforce BoundedExecutor max usage of threads

The prior implementation could have created more than maxThreads that were very short lived.

erichwang committed:
Report if BoundedExecutor reaches a corrupt state

electrum committed:
Remove BoundedThreadPool

The current implementation seems to have weird scheduling latencies
and potentially other issues.

martint committed:
[maven-release-plugin] prepare release 0.128

martint committed:
[maven-release-plugin] prepare for next development iteration

dain committed:
Ignore EOFException while writing json http response

Commit:
Add truststore configuration

Commit:
Expose EventTypeMetadata and EventFieldMetadata

This is required for non-JSON serialization strategies to gain access
to type metadata without having to inspect JsonSerializer.

Commit:
Remove no-name requirement for nested event data

It's useful for all event classes to have names so that non-JSON
serialization strategies can access type metadata without having to
inspect JsonSerializers.

dain commented:
@martint is this something you want to review?

nileema committed:
Add CpuTimer to stats

nileema committed:
Update release notes

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -285,15 +285,20 @@ public boolean process(Node node)
-        return getQuantilesUpperBound(quantiles);
+        List<Long> accumulator = new ArrayList<Long>(quantiles.size());
+
+        for (Double q : quantiles) {
```
this can be `double q`

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -371,6 +376,91 @@ public boolean process(Node node)
         return builder.build();
     }
 
+    public double getCDF(long x) {
```
Don't abbreviate variable names.

DanielTing replied:
CDF in stats nomenclature is almost like CPU in cs. Is that acceptable use?
The alternative is getCumulativeDistributionFunction(long x)

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -371,6 +376,91 @@ public boolean process(Node node)
     }
 
+    public double getCDF(long x) {
+        final ImmutableList.Builder<Double> builder = ImmutableList.builder();
```
We generally don't use `final` for local variables

martint replied:
Is this being used as a way to return a single value from the callback traversal? It'd be cleaner to either:
1) Introduce a mutable value holder that the callback sets.
2) Pull the callback code into a named class, add a member variable and accessor.

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -371,6 +376,91 @@ public boolean process(Node node)
+        final long value = x;
+
+        inOrderTraversal(root, new Callback() {
+            protected double sum = 0.0;
```
make this `private`

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -285,15 +285,20 @@ public boolean process(Node node)
 
```
Please, reformat according to https://github.com/airlift/codestyle

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -371,6 +376,91 @@ public boolean process(Node node)
+                } else {
+                    // interpolate
+                    double ratio = (value - lower) / (double) (upper - lower);
+                    if (node.right == null && node.left != null) {
```
Can you add a comment explaining this case?

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -882,6 +972,39 @@ private boolean postOrderTraversal(Node node, Callback callback, TraversalOrder
         return callback.process(node);
     }
 
+    protected boolean inOrderTraversal(Node node, Callback callback, TraversalOrder order)
```
This should be private. This class is not meant to be extended, so there's not benefit in making it protected.

martint commented on stats/src/main/java/io/airlift/stats/QuantileDigest.java:
```diff
@@ -371,6 +376,91 @@ public boolean process(Node node)
+        }
+    }
+
+    public long getQuantileFromCDF(double p) {
```
This needs an explanation about how the algorithm works.

martint commented:
Can you add some tests?

electrum committed:
Add FileBodyGenerator for HTTP client

Commit:
Change dataType, nestedType and containerType to be Optional

We now use Optional.empty() rather than null to signal the lack of
these properties.

Commit:
Update release notes for 0.129

electrum committed:
Update CHANGES and NEWS

cberner committed:
Update to Jetty 9.3.11.M0

This fixes a race when using async responses

martint committed:
[maven-release-plugin] prepare for next development iteration

martint committed:
[maven-release-plugin] prepare release 0.129

Commit:
Add getField method to EventTypeMetadata

Also changed member variable to be of type SortedMap instead of List.

electrum committed:
Move Jetty and Jersey dependencies from Airbase

electrum committed:
Update CHANGES and NEWS

electrum committed:
[maven-release-plugin] prepare release 0.130

electrum committed:
[maven-release-plugin] prepare for next development iteration

electrum committed:
Statically configure logging for Jetty HTTP client

This removes the need for the logging properties file.
It also suppresses the "Logging initialized message.

kaschaeffer commented:
ping @DanielTing can you address comments?  Would be awesome to get this merge-ready.

Commit:
Fix documentation and codestyle for quantile CDF

Commit:
remove testing code

DanielTing commented:
@martint What tests did you have in mind? The usual quantile ones generally cover the modifications, and getCDF is now private.

DanielTing commented:
Also, tomorrow is my last day, so if there are any changes requested please let me know now!

Commit:
Additional comments on improved quantile algorithm

kaschaeffer commented:
@DanielTing should I take this over since you're no longer at fb?  I'd still like to get this merged in eventually.

kaschaeffer commented:
Taking this over.
